File: day12/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var start, end Pos
	hasStart, hasEnd := false, false

	var starts2 []Pos

	var grid [][]byte

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := []byte(scanner.Text())
		for x, i := range line {
			switch i {
			case 'S':
				start = Pos{
					y: len(grid),
					x: x,
				}
				hasStart = true
				line[x] = 'a'
				starts2 = append(starts2, Pos{
					y: len(grid),
					x: x,
				})
			case 'E':
				end = Pos{
					y: len(grid),
					x: x,
				}
				hasEnd = true
				line[x] = 'z'
			case 'a':
				starts2 = append(starts2, Pos{
					y: len(grid),
					x: x,
				})
			}
		}
		grid = append(grid, line)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	if !hasStart || !hasEnd {
		log.Fatalln("Missing start or end")
	}

	g := NewGrid(grid)
	{
		_, steps := AStarSearch([]Pos{start}, end, g.neighbors, manhattanDistance)
		fmt.Println("Part 1:", steps)
	}
	{
		_, steps := AStarSearch(starts2, end, g.neighbors, manhattanDistance)
		fmt.Println("Part 2:", steps)
	}
}

type (
	Pos struct {
		y, x int
	}

	Grid struct {
		h    int
		w    int
		grid [][]byte
	}
)

func NewGrid(grid [][]byte) *Grid {
	return &Grid{
		h:    len(grid),
		w:    len(grid[0]),
		grid: grid,
	}
}

func (g Grid) inBounds(k Pos) bool {
	return k.x >= 0 && k.x < g.w && k.y >= 0 && k.y < g.h
}

func (g Grid) get(k Pos) (byte, bool) {
	if !g.inBounds(k) {
		return 0, false
	}
	return g.grid[k.y][k.x], true
}

var (
	dirDeltas = []Pos{
		{-1, 0},
		{0, 1},
		{1, 0},
		{0, -1},
	}
)

func (g Grid) neighbors(k Pos) []AStarEdge[Pos] {
	limit := g.grid[k.y][k.x] + 1
	var e []AStarEdge[Pos]
	for _, i := range dirDeltas {
		k := Pos{
			y: k.y + i.y,
			x: k.x + i.x,
		}
		v, ok := g.get(k)
		if !ok || v > limit {
			continue
		}
		e = append(e, AStarEdge[Pos]{
			Key: k,
			DG:  1,
		})
	}
	return e
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func manhattanDistance(a, b Pos) int {
	return abs(a.x-b.x) + abs(a.y-b.y)
}

File: day12/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

mod astar;

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut start = None;
    let mut end = None;
    let mut starts2 = Vec::new();

    let mut grid = Vec::new();

    for line in reader.lines() {
        grid.push(
            line?
                .bytes()
                .enumerate()
                .map(|(x, i)| {
                    let pos = Pos { y: grid.len(), x };
                    match i {
                        b'S' => {
                            start = Some(pos);
                            starts2.push(pos);
                            b'a'
                        }
                        b'E' => {
                            end = Some(pos);
                            b'z'
                        }
                        b'a' => {
                            starts2.push(pos);
                            b'a'
                        }
                        c => c,
                    }
                })
                .collect::<Vec<_>>(),
        );
    }

    let (start, end) = match (start, end) {
        (Some(start), Some(end)) => (start, end),
        _ => return Err("Missing start or end".into()),
    };

    let grid = Grid::new(grid);

    if let Some((_, steps)) = astar::search(&vec![start], &end, &grid, manhattan_distance) {
        println!("Part 1: {}", steps);
    }

    if let Some((_, steps)) = astar::search(&starts2, &end, &grid, manhattan_distance) {
        println!("Part 2: {}", steps);
    }

    Ok(())
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
struct Pos {
    y: usize,
    x: usize,
}

struct Grid {
    h: usize,
    w: usize,
    grid: Vec<Vec<u8>>,
}

const DIR_DELTAS: [(isize, isize); 4] = [(-1, 0), (0, 1), (1, 0), (0, -1)];

impl Grid {
    fn new(grid: Vec<Vec<u8>>) -> Self {
        Self {
            h: grid.len(),
            w: grid[0].len(),
            grid,
        }
    }

    fn in_bounds(&self, k: &Pos) -> bool {
        k.x < self.w && k.y < self.h
    }

    fn get(&self, k: &Pos) -> Option<u8> {
        if self.in_bounds(k) {
            Some(self.grid[k.y][k.x])
        } else {
            None
        }
    }
}

impl astar::Neighborer<Pos> for Grid {
    fn neighbors(&self, k: &Pos) -> Vec<astar::Edge<Pos>> {
        let limit = self.grid[k.y][k.x] + 1;
        let mut e = Vec::new();
        for (dy, dx) in DIR_DELTAS {
            let k = Pos {
                y: k.y.wrapping_add_signed(dy),
                x: k.x.wrapping_add_signed(dx),
            };
            if let Some(v) = self.get(&k) {
                if v <= limit {
                    e.push(astar::Edge { value: k, dg: 1 })
                }
            }
        }
        e
    }
}

fn manhattan_distance(a: &Pos, b: &Pos) -> usize {
    a.x.abs_diff(b.x) + a.y.abs_diff(b.y)
}
